### PR TITLE
🐛 Add a preflight check to ensure that custom providers have a fetch config

### PR DESCRIPTION
**Pull request by Fedosin**
**What this PR does / why we need it**:

Currently users don't have to specify fetch configs for well-known providers, but for custom providers it is mandatory. Without it provider installation fails in runtime with an obscure error message.

This PR adds a preflight check to ensure that custom providers always have a non-empty fetch config.

### CODE REVIEW

File: internal/controller/preflight_checks_test.go [hunk 1]
```diff
@@ -531,6 +531,102 @@ func TestPreflightChecks(t *testing.T) {
...
+				Severity: clusterv1.ConditionSeverityError,
+				Message:  "Either Selector or URL must be provided for a not predefined provider",
+				Status:   corev1.ConditionFalse,
+			},
+			providerList: &genericprovider.InfrastructureProviderListWrapper{
+				InfrastructureProviderList: &operatorv1.InfrastructureProviderList{},
+			},
+		},
+		{
+			name:          "custom Core Provider with empty fetch config, preflight check failed",
```
Comment (furkatgofurov7) [inline internal/controller/preflight_checks_test.go:597]: this and previous tests are very similar: how about adding a test with empty fetch config Selector & URL?

```suggestion
		{
			name:          "custom Core Provider with empty selector in fetch config, preflight check failed",
			expectedError: true,
			providers: []genericprovider.GenericProvider{
				&genericprovider.CoreProviderWrapper{
					CoreProvider: &operatorv1.CoreProvider{
						ObjectMeta: metav1.ObjectMeta{
							Name:      "my-custom-cluster-api",
							Namespace: namespaceName1,
						},
						TypeMeta: metav1.TypeMeta{
							Kind:       "CoreProvider",
							APIVersion: "operator.cluster.x-k8s.io/v1alpha1",
						},
						Spec: operatorv1.CoreProviderSpec{
							ProviderSpec: operatorv1.ProviderSpec{
								Version:     "v1.0.0",
								FetchConfig: &operatorv1.FetchConfiguration{
									Selector: &metav1.LabelSelector{},
								},
							},
						},
					},
				},
			},
			expectedCondition: clusterv1.Condition{
				Type:     operatorv1.PreflightCheckCondition,
				Reason:   operatorv1.FetchConfigValidationErrorReason,
				Severity: clusterv1.ConditionSeverityError,
				Message:  "Either Selector or URL must be provided for a not predefined provider",
				Status:   corev1.ConditionFalse,
			},
			providerList: &genericprovider.CoreProviderListWrapper{
				CoreProviderList: &operatorv1.CoreProviderList{},
			},
		},
		{
			name:          "custom Core Provider with empty URL in fetch config, preflight check failed",
			expectedError: true,
			providers: []genericprovider.GenericProvider{
				&genericprovider.CoreProviderWrapper{
					CoreProvider: &operatorv1.CoreProvider{
						ObjectMeta: metav1.ObjectMeta{
							Name:      "my-custom-cluster-api",
							Namespace: namespaceName1,
						},
						TypeMeta: metav1.TypeMeta{
							Kind:       "CoreProvider",
							APIVersion: "operator.cluster.x-k8s.io/v1alpha1",
						},
						Spec: operatorv1.CoreProviderSpec{
							ProviderSpec: operatorv1.ProviderSpec{
								Version:     "v1.0.0",
								FetchConfig: &operatorv1.FetchConfiguration{
									URL: "",
								},
							},
						},
					},
				},
			},
			expectedCondition: clusterv1.Condition{
				Type:     operatorv1.PreflightCheckCondition,
				Reason:   operatorv1.FetchConfigValidationErrorReason,
				Severity: clusterv1.ConditionSeverityError,
				Message:  "Either Selector or URL must be provided for a not predefined provider",
				Status:   corev1.ConditionFalse,
			},
			providerList: &genericprovider.CoreProviderListWrapper{
				CoreProviderList: &operatorv1.CoreProviderList{},
			},
		},
```

**Comment by Fedosin** [comment]
@furkatgofurov7 should be it :)

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

/cc @alexander-demicev

**Review by alexander-demicev — APPROVED** [review]
/approve
